From 6a1f217d30f87311a51aee7fc48e19a4eca00ca7 Mon Sep 17 00:00:00 2001
From: Peter Nelson <peter1138@openttd.org>
Date: Sun, 24 Mar 2019 00:12:31 +0000
Subject: [PATCH 0477/1053] Fix #7400: Water class for tree tiles was not
 converted for old saves preventing industry creation.

As the information is always available from the tree ground type, unconditionally
update the map array for tree tiles.
---
 src/saveload/afterload.cpp | 7 +++++++
 src/tree_map.h             | 1 +
 2 files changed, 8 insertions(+)

diff --git a/src/saveload/afterload.cpp b/src/saveload/afterload.cpp
index 0b570966dc..96def205c5 100644
--- a/src/saveload/afterload.cpp
+++ b/src/saveload/afterload.cpp
@@ -3080,6 +3080,13 @@ bool AfterLoadGame()
 		}
 	}
 
+	{
+		/* Update water class for trees for all current savegame versions. */
+		for (TileIndex t = 0; t < map_size; t++) {
+			if (IsTileType(t, MP_TREES)) SetWaterClass(t, GetTreeGround(t) == TREE_GROUND_SHORE ? WATER_CLASS_SEA : WATER_CLASS_INVALID);
+		}
+	}
+
 	/* Station acceptance is some kind of cache */
 	if (IsSavegameVersionBefore(SLV_127)) {
 		Station *st;
diff --git a/src/tree_map.h b/src/tree_map.h
index df9fd441cc..bd1567b54c 100644
--- a/src/tree_map.h
+++ b/src/tree_map.h
@@ -277,6 +277,7 @@ static inline void MakeTree(TileIndex t, TreeType type, uint count, uint growth,
 {
 	SetTileType(t, MP_TREES);
 	SetTileOwner(t, OWNER_NONE);
+	SetWaterClass(t, ground == TREE_GROUND_SHORE ? WATER_CLASS_SEA : WATER_CLASS_INVALID);
 	_m[t].m2 = ground << 6 | density << 4 | 0;
 	_m[t].m3 = type;
 	_m[t].m4 = 0 << 5 | 0 << 2;
-- 
2.55.0

